aten/src/ATen/native/cpu/BinaryOpsKernel.cpp

```diff
@@ -314,12 +314,21 @@ void bitwise_xor_kernel(TensorIteratorBase& iter) {
 
 void lshift_kernel(TensorIteratorBase& iter) {
   AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "lshift_cpu", [&]() {
+    using vec_t = Vectorized<scalar_t>;
+    constexpr scalar_t shift = sizeof(scalar_t) * CHAR_BIT;
+    const vec_t zero_vec(0);
+    const vec_t shift_vec(shift);
     cpu_kernel_vec(iter,
         [](scalar_t a, scalar_t b) -> scalar_t {
-          return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+          if ((b >= 0) && (b < shift)) {
+            return static_cast<std::make_unsigned_t<scalar_t>>(a) << b;
+          }
+          return 0;
         },
-        [](Vectorized<scalar_t> a, Vectorized<scalar_t> b) {
-            return a << b;
+        [&zero_vec, &shift_vec](vec_t a, vec_t b) -> vec_t {
+          auto mask = (b >= zero_vec) & (b < shift_vec);
+          a = vec_t::blendv(zero_vec, a, mask);
+          return a << b;
```

Comment:
May I know what is the problem for vectorized implementation here? The `_mm256_srav_epi32` and `_mm256_sllv_epi32` etc. should already do the right things and we don't have to do extra masking here, right?

Reply:
What do you mean by "the right thing"? These are UB from the standard iirc.

Reply:
I mean the semantics of the intrinsics/instructions already guarantee the expected behavior, e.g., when shifting left with outside-range count, zero is returned and when shifting right, sign flag is preserved, no need to do line 329-330 (and similarly for right shift). Or did I miss anything here?

Reply:
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_srav_epi32
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_sllv_epi32

Reply:
Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

Reply:
The nightly macos build is compiled w/o AVX2 and it fails the new test. So a fix is necessary for other CPUs.

An alternate solution, which assumes the SIMD implementations of `operator<<` and `operator>>` are correct, would be to fix the `operator<<` and `operator>>` implementations in `vec_base.h` for when SIMD implementation is not supported. I will try this.

Reply:
Turns out there was a bug in `vec256` for the [`operator>>` implementation for `int64`](https://github.com/pytorch/pytorch/blob/ba4fb9b6ad0b3a06ae1b15a170cea6f4223bda79/aten/src/ATen/cpu/vec/vec256/vec256_int.h#L1482-L1497). Since there is no shift right arithmetic vector instruction, the implementation emulated the instruction. However the implementation did not work if the shift value was out-of-limit (<0 or >64). When the shift value was out-of-limit, subtracting the shift value from 64 to compute the sign shift will result in a bad sign shift value and bad sign extension value for negative input.

I have added a new commit to the PR which fixes this.

Reply:
> Thanks for the details. Is tha ttrue for all CPU though? Including older Intel hardware that might implement an older version of the intrinsics? And non-Intel hardware?

The behavior should be consistent on x86 CPUs in general. In general, I think it is better to address the problem in the vectorized implementation to avoid unnecessary guards where possible.